### ISSUE TITLE
Nested record setters disturb type inference for ad-hoc polymorphic expressions

### ISSUE THREAD

**Issue by jberthold-da** [digital-asset/daml#9717]
The following program with nested record setters and a `HasDefault` class fails to compile:
```
data One = One with Two : Two
data Two = Two with val : Int

class HasDefault a where def : a
instance HasDefault Int where def = 0
instance HasDefault One where def = One def
instance HasDefault Two where def = Two def

works   = (def : One) with two.val = 42
doznwok : One
doznwok = def with two.val = 42
```
Irritatingly, the error (ambiguous type var from `setField`) reports only one existing instance.
/home/jost/workspace/code/solution-spider2/app/daml-model/main/DA/ASX/Main/Conversions/Test.daml:69:11: error:
```
    • Ambiguous type variable ‘r0’ arising from a use of ‘getField’
      prevents the constraint ‘(DA.Internal.Record.HasField
                                  "two" r0 Two)’ from being solved.
      Probable fix: use a type annotation to specify what ‘r0’ should be.
      These potential instance exist:
        instance DA.Internal.Record.HasField "two" One Two
          -- Defined in ‘Test’
    • In the third argument of ‘setField’, namely ‘getField @"two" def’
      In the second argument of ‘setField’, namely
        ‘setField @"val" 42 getField @"two" def’
      In the expression:
        setField @"two" setField @"val" 42 getField @"two" def def
```
Judging from the last line (and imagining some parentheses where they would make sense), the desugaring seems to duplicate the `def` before `with`, but loses its type context.

**Comment by cocreature** [comment]
@robin-da could you please take a look at this? We should be desugaring that into something like `let x = def in setField @"two" (setField @"val" 42 (getField @"two" x)) x`

**Comment by jberthold-da** [comment]
I actually wrote the above by hand, and it has the same problem (`x` being `def`, polymorphic)

**Comment by cocreature** [comment]
standalone example that does work
```
module Main where

data One = One with two : Two
data Two = Two with val : Int

class HasDefault a where def : a
instance HasDefault Int where def = 0
instance HasDefault One where def = One def
instance HasDefault Two where def = Two def

y : One
y = let x = def in setField @"two" (setField @"val" 42 (getField @"two" x)) x
```

**Comment by robin-da** [comment]
Fixed with a3363630202aad96661e3d03937b85e58e0db3a5.
